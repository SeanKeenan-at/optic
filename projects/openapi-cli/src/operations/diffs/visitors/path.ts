import { OpenAPIV3 } from '../../../specs';
import { OperationDiffResult, OperationDiffResultKind } from '../result';
import { PathComponents, PathComponent } from '../..';
import { Option } from 'ts-results';

export function* visitPath(
  path: string,
  specOption: Option<OpenAPIV3.PathItemObject>,
  context: { pathPattern: Option<string> }
): IterableIterator<OperationDiffResult> {
  let parameterComponents = PathComponents.fromPath(path).filter(
    PathComponent.isTemplate
  );

  if (specOption.none) {
    yield {
      kind: OperationDiffResultKind.UnmatchedPath,
      subject: path,
    };

    for (let parameterComponent of parameterComponents) {
      yield {
        kind: OperationDiffResultKind.UnmatchedPathParameter,
        subject: parameterComponent.name,
      };
    }
  }

  // TODO: match path's parameters against those documented
  // const spec = specOption.unwrap()
  // spec.parameters
}
