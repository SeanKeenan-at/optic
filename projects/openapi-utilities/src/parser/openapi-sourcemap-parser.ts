import $RefParser from "@apidevtools/json-schema-ref-parser";
// @ts-ignore
import * as $RefParserOptions from "@apidevtools/json-schema-ref-parser/lib/options";
import * as YAML from "yaml-ast-parser";
import * as fs from "fs-extra";
import { YAMLMapping, YAMLNode, YAMLSequence } from "yaml-ast-parser";
// @ts-ignore
import { dereference } from "./insourced-dereference";
import * as pointer from "json-ptr";
import path from "path";
import fetch from "node-fetch";
import { OpenAPIV3 } from "openapi-types";
import sortBy from "lodash.sortby";

export type ParseOpenAPIResult = {
  jsonLike: OpenAPIV3.Document;
  sourcemap: JsonSchemaSourcemapOutput;
};

export async function parseOpenAPIWithSourcemap(
  path: string
): Promise<ParseOpenAPIResult> {
  const resolver = new $RefParser();

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path, {
    resolve: {
      http: {
        headers: {
          accept: "*/*",
        },
      },
    },
  });

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map((filePath, index) => sourcemap.addFileIfMissing(filePath, index))
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: path },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

export async function parseOpenAPIFromRepoWithSourcemap(
  name: string,
  repoPath: string,
  branch: string
): Promise<ParseOpenAPIResult> {
  const newGitBranchResolver = require("./git-branch-file-resolver.js");

  const inGitResolver = newGitBranchResolver(repoPath, branch);
  const resolver = new $RefParser();
  const fileName = path.join(repoPath, name);

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(fileName, {
    resolve: { file: inGitResolver },
  });

  // parse all asts
  await Promise.all(
    resolverResults.paths().map(async (filePath, index) => {
      return await sourcemap.addFileIfMissingFromContents(
        filePath,
        await inGitResolver.read({ url: filePath }),
        index
      );
    })
  );

  dereference(
    resolver,
    {
      ...$RefParserOptions.defaults,
      path: fileName,
      resolve: { file: inGitResolver },
    },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

type JsonPath = string;
type FileReference = number;

type DerefToSource = [JsonPath, LocationRecord];

export interface JsonSchemaSourcemapOutput {
  files: Array<{
    path: string;
    index: number;
  }>;
  map: DerefToSource[];
}

export class JsonSchemaSourcemap {
  private _files: Array<{
    path: string;
    index: number;
    ast: YAMLNode;
  }> = [];

  private _mappings: Array<DerefToSource> = [];

  async addFileIfMissing(filePath: string, fileIndex: number) {
    if (filePath.startsWith("http")) {
      const response = await fetch(filePath);
      const asText = await response.text();

      const yamlAst: YAMLNode = YAML.safeLoad(asText);

      this._files.push({
        path: filePath,
        index: fileIndex,
        ast: yamlAst,
      });
    } else {
      if (!this._files.find((i) => i.path === filePath)) {
        // add the ast to the cache
        const yamlAst: YAMLNode = YAML.safeLoad(
          (await fs.readFile(filePath)).toString()
        );

        this._files.push({
          path: filePath,
          index: fileIndex,
          ast: yamlAst,
        });
      }
    }
  }

  async addFileIfMissingFromContents(
    filePath: string,
    contents: string,
    fileIndex: number
  ) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(contents);

      this._files.push({
        path: filePath,
        index: fileIndex,
        ast: yamlAst,
      });
    }
  }

  log(path: string, pathFromRoot: string) {
    // this seems to assume that paths will be in order, why not check for equality?
    const thisFile = this._files.find((i) => path.startsWith(i.path));
    if (thisFile) {
      const jsonPointer = path.split(thisFile.path)[1].substring(1) || "/";
      const sourceMapping = resolveJsonPointerInYamlAst(
        thisFile.ast,
        jsonPointer,
        thisFile.index
      );
      if (sourceMapping) {
        this._mappings.push([pathFromRoot, sourceMapping]);
      }
    }
  }

  public serialize(): JsonSchemaSourcemapOutput {
    return {
      files: sortBy(
        this._files.map((i) => ({ path: i.path, index: i.index })),
        "index"
      ),
      map: this._mappings,
    };
  }
}

export function resolveJsonPointerInYamlAst(
  node: YAMLNode,
  jsonPointer: string,
  file: number
): LocationRecord | undefined {
  const decoded = pointer.decodePointer(jsonPointer);

  const isEmpty =
    decoded.length === 0 || (decoded.length === 1 && decoded[0] === "");

  if (isEmpty)
    return { node: [node.startPosition, node.endPosition], file: file };

  const found: YAMLNode | undefined = decoded.reduce((current, path) => {
    if (!current) return undefined;
    const node: YAMLNode = current.key ? current.value : current;
    const isNumericalKey =
      !isNaN(Number(path)) && (node as any).hasOwnProperty("items");

    if (isNumericalKey) {
      return (node as YAMLSequence).items[Number(path)];
    } else {
      const field = node.mappings.find(
        (i: YAMLMapping) => i.key.value === path
      );
      return field;
    }
  }, node as YAMLNode | undefined);

  if (found) {
    if (found.key) {
      // is a field
      return {
        key: [found.key.startPosition, found.key.endPosition],
        value: [found.value.startPosition, found.value.endPosition],
        node: [found.startPosition, found.endPosition],
        file: file,
      };
    } else {
      return { node: [found.startPosition, found.endPosition], file: file };
    }
  }
}

type AstLocation = [number, number];
interface LocationRecord {
  key?: AstLocation;
  value?: AstLocation;
  node: AstLocation;
  file: number;
}
