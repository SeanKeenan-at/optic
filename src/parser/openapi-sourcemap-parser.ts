import $RefParser from "@apidevtools/json-schema-ref-parser";
// @ts-ignore
import * as $RefParserOptions from "@apidevtools/json-schema-ref-parser/lib/options";
import * as YAML from "yaml-ast-parser";
import * as fs from "fs-extra";
import { YAMLMapping, YAMLNode, YAMLSequence } from "yaml-ast-parser";
// @ts-ignore
import { dereference } from "./insourced-dereference";
import * as pointer from "json-ptr";
const newGitBranchResolver = require("./git-branch-file-resolver")
import path from 'path'


export async function parseOpenAPIWithSourcemap(path: string) {
  const resolver = new $RefParser();

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(path);

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map((filePath) => sourcemap.addFileIfMissing(filePath))
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: path },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

export async function parseOpenAPIFromRepoWithSourcemap(name: string, repoPath: string, branch: string) {
  const inGitResolver = newGitBranchResolver(repoPath, branch)
  const resolver = new $RefParser();
  const fileName = path.join(repoPath, name)

  const sourcemap = new JsonSchemaSourcemap();
  const resolverResults: $RefParser.$Refs = await resolver.resolve(fileName, {resolve: {file: inGitResolver}});

  // parse all asts
  await Promise.all(
    resolverResults
      .paths()
      .map(async (filePath) => {
        return await sourcemap.addFileIfMissingFromContents(filePath, await inGitResolver.read({url: filePath}))
      })
  );

  dereference(
    resolver,
    { ...$RefParserOptions.defaults, path: fileName, resolve: {file: inGitResolver} },
    sourcemap
  );

  return { jsonLike: resolver.schema as any, sourcemap: sourcemap.serialize() };
}

type JsonPath = string;
type FileReference = number;

type DerefToSource = [JsonPath, LocationRecord];

export interface JsonSchemaSourcemapOutput {
  files: Array<{
    path: string;
    index: number;
  }>;
  map: DerefToSource[];
}

export class JsonSchemaSourcemap {
  private _files: Array<{
    path: string;
    index: number;
    ast: YAMLNode;
  }> = [];

  private _mappings: Array<DerefToSource> = [];

  async addFileIfMissing(filePath: string) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(
        (await fs.readFile(filePath)).toString()
      );

      this._files.push({
        path: filePath,
        index: this._files.length,
        ast: yamlAst,
      });
    }
  }

  async addFileIfMissingFromContents(filePath: string, contents: string) {
    if (!this._files.find((i) => i.path === filePath)) {
      // add the ast to the cache
      const yamlAst: YAMLNode = YAML.safeLoad(
        contents
      );

      this._files.push({
        path: filePath,
        index: this._files.length,
        ast: yamlAst,
      });
    }
  }

  log(path: string, pathFromRoot: string) {
    const thisFile = this._files.find((i) => path.startsWith(i.path));
    if (thisFile) {
      const jsonPointer = path.split(thisFile.path)[1].substring(1) || "/";
      // @todo remove this try catch, we want errors, but this is going to help us dev
      try {
        const sourceMapping = resolveJsonPointerInYamlAst(
          thisFile.ast,
          jsonPointer,
          thisFile.index
        );
        if (sourceMapping) {
          this._mappings.push([pathFromRoot, sourceMapping]);
        }
      } catch (e) {
        console.error("source mapping failure ", e)
      }
    }
  }

  public serialize(): JsonSchemaSourcemapOutput {
    return {
      files: this._files.map((i) => ({ path: i.path, index: i.index })),
      map: this._mappings,
    };
  }
}

export function resolveJsonPointerInYamlAst(
  node: YAMLNode,
  jsonPointer: string,
  file: number
): LocationRecord | undefined {
  const decoded = pointer.decodePointer(jsonPointer);

  const isEmpty =
    decoded.length === 0 || (decoded.length === 1 && decoded[0] === "");

  if (isEmpty) return { n: [node.startPosition, node.endPosition], f: file };

  const found: YAMLNode | undefined = decoded.reduce((current, path) => {
    const isFieldKey = isNaN(Number(path));
    if (!current) return undefined;

    const node: YAMLNode = current.key ? current.value : current;

    if (isFieldKey) {
      const field = node.mappings.find(
        (i: YAMLMapping) => i.key.value === path
      );
      return field;
    } else {
      // is number
      return (node as YAMLSequence).items[Number(path)];
    }
  }, node as YAMLNode | undefined);

  if (found) {
    if (found.key) {
      // is a field
      return {
        k: [found.key.startPosition, found.key.endPosition],
        v: [found.value.startPosition, found.value.endPosition],
        n: [found.startPosition, found.endPosition],
        f: file,
      };
    } else {
      return { n: [found.startPosition, found.endPosition], f: file };
    }
  }
}

type AstLocation = [number, number];
interface LocationRecord {
  k?: AstLocation;
  v?: AstLocation;
  n: AstLocation;
  f: number;
}
