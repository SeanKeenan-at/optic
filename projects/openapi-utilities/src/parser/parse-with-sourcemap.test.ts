import tap from 'tap'
import {parseOpenAPIWithSourcemap} from "./openapi-sourcemap-parser";
import path from "path";

tap.test("can parse an OpenAPI spec with external references", async () => {

  const results = await parseOpenAPIWithSourcemap(
    path.resolve(
      path.join(__dirname, "../../inputs/openapi3-with-references/external-multiple.yaml")
    )
  );

  tap.matchSnapshot(results);

});
