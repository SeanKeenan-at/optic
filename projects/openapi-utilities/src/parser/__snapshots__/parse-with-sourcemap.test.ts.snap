// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can parse an OpenAPI spec with external references 1`] = `
Object {
  "jsonLike": Object {
    "example": Object {
      "token": "11111111",
      "user": Object {
        "name": "Homer",
      },
    },
    "properties": Object {
      "token": Object {
        "type": "string",
      },
      "user": Object {
        "example": Object {
          "name": "Homer",
        },
        "properties": Object {
          "name": Object {
            "type": "string",
          },
        },
        "required": Array [
          "name",
        ],
        "type": "object",
      },
    },
    "required": Array [
      "user",
      "token",
    ],
    "type": "object",
  },
  "sourcemap": Object {
    "files": Array [
      Object {
        "index": 0,
        "path": "/inputs/openapi3-with-references/external-multiple.yaml",
      },
      Object {
        "index": 1,
        "path": "/inputs/openapi3-with-references/definitions.yaml",
      },
    ],
    "map": Array [
      Array [
        "#",
        Object {
          "file": 0,
          "node": Array [
            0,
            198,
          ],
        },
      ],
      Array [
        "#/type",
        Object {
          "file": 0,
          "key": Array [
            0,
            4,
          ],
          "node": Array [
            0,
            12,
          ],
          "value": Array [
            6,
            12,
          ],
        },
      ],
      Array [
        "#/required",
        Object {
          "file": 0,
          "key": Array [
            13,
            21,
          ],
          "node": Array [
            13,
            36,
          ],
          "value": Array [
            23,
            36,
          ],
        },
      ],
      Array [
        "#/required/0",
        Object {
          "file": 0,
          "node": Array [
            24,
            28,
          ],
        },
      ],
      Array [
        "#/required/1",
        Object {
          "file": 0,
          "node": Array [
            30,
            35,
          ],
        },
      ],
      Array [
        "#/properties",
        Object {
          "file": 0,
          "key": Array [
            37,
            47,
          ],
          "node": Array [
            37,
            117,
          ],
          "value": Array [
            51,
            117,
          ],
        },
      ],
      Array [
        "#/properties/token",
        Object {
          "file": 0,
          "key": Array [
            51,
            56,
          ],
          "node": Array [
            51,
            74,
          ],
          "value": Array [
            62,
            74,
          ],
        },
      ],
      Array [
        "#/properties/token/type",
        Object {
          "file": 0,
          "key": Array [
            62,
            66,
          ],
          "node": Array [
            62,
            74,
          ],
          "value": Array [
            68,
            74,
          ],
        },
      ],
      Array [
        "#/properties/user",
        Object {
          "file": 1,
          "key": Array [
            0,
            4,
          ],
          "node": Array [
            0,
            111,
          ],
          "value": Array [
            8,
            111,
          ],
        },
      ],
      Array [
        "#/properties/user/type",
        Object {
          "file": 1,
          "key": Array [
            8,
            12,
          ],
          "node": Array [
            8,
            20,
          ],
          "value": Array [
            14,
            20,
          ],
        },
      ],
      Array [
        "#/properties/user/required",
        Object {
          "file": 1,
          "key": Array [
            23,
            31,
          ],
          "node": Array [
            23,
            39,
          ],
          "value": Array [
            33,
            39,
          ],
        },
      ],
      Array [
        "#/properties/user/required/0",
        Object {
          "file": 1,
          "node": Array [
            34,
            38,
          ],
        },
      ],
      Array [
        "#/properties/user/properties",
        Object {
          "file": 1,
          "key": Array [
            42,
            52,
          ],
          "node": Array [
            42,
            82,
          ],
          "value": Array [
            58,
            82,
          ],
        },
      ],
      Array [
        "#/properties/user/properties/name",
        Object {
          "file": 1,
          "key": Array [
            58,
            62,
          ],
          "node": Array [
            58,
            82,
          ],
          "value": Array [
            70,
            82,
          ],
        },
      ],
      Array [
        "#/properties/user/properties/name/type",
        Object {
          "file": 1,
          "key": Array [
            70,
            74,
          ],
          "node": Array [
            70,
            82,
          ],
          "value": Array [
            76,
            82,
          ],
        },
      ],
      Array [
        "#/properties/user/example",
        Object {
          "file": 1,
          "key": Array [
            85,
            92,
          ],
          "node": Array [
            85,
            111,
          ],
          "value": Array [
            98,
            111,
          ],
        },
      ],
      Array [
        "#/properties/user/example/name",
        Object {
          "file": 1,
          "key": Array [
            98,
            102,
          ],
          "node": Array [
            98,
            111,
          ],
          "value": Array [
            104,
            111,
          ],
        },
      ],
      Array [
        "#/example",
        Object {
          "file": 0,
          "key": Array [
            118,
            125,
          ],
          "node": Array [
            118,
            197,
          ],
          "value": Array [
            129,
            197,
          ],
        },
      ],
      Array [
        "#/example/token",
        Object {
          "file": 0,
          "key": Array [
            129,
            134,
          ],
          "node": Array [
            129,
            146,
          ],
          "value": Array [
            136,
            146,
          ],
        },
      ],
      Array [
        "#/example/user",
        Object {
          "file": 1,
          "key": Array [
            85,
            92,
          ],
          "node": Array [
            85,
            111,
          ],
          "value": Array [
            98,
            111,
          ],
        },
      ],
    ],
  },
}
`;
